pkg/integrations/v2/app_o11y_receiver/sourcemaps/sourcemaps.go

```diff
@@ -0,0 +1,331 @@
+		}
+		resolvedSourceMapURL = base.ResolveReference(relative).String()
+		level.Debug(store.l).Log("msg", "resolved absolute soure map url", "url", sourceURL, "sourceMapURL", resolvedSourceMapURL)
+	}
+	level.Debug(store.l).Log("msg", "attempting to download sourcemap file", "url", resolvedSourceMapURL)
+	result, err = store.downloadFileContents(resolvedSourceMapURL)
+	if err != nil {
+		level.Debug(store.l).Log("failed to download source map file", "url", resolvedSourceMapURL, "err", err)
+		return nil, "", err
+	}
+	return result, resolvedSourceMapURL, nil
+}
+
+func (store *RealSourceMapStore) getSourceMapFromFileSystem(sourceURL string, release string, fileconf config.SourceMapFileLocation) (content []byte, sourceMapURL string, err error) {
+	if len(sourceURL) == 0 || !strings.HasPrefix(sourceURL, fileconf.MinifiedPathPrefix) || strings.HasSuffix(sourceURL, "/") {
+		return nil, "", nil
+	}
+	pathParts := []string{strings.Replace(fileconf.Path, "{RELEASE}", cleanForFilePath(release), 1)}
+	for _, part := range strings.Split(strings.TrimPrefix(sourceURL, fileconf.MinifiedPathPrefix), "/") {
+		if len(part) > 0 && part != "." && part != ".." {
+			pathParts = append(pathParts, part)
+		}
+	}
+	mapFilePath := filepath.Join(pathParts...) + ".map"
+
+	if _, err := store.fileService.Stat(mapFilePath); err != nil {
+		store.metrics.fileReads.WithLabelValues(getOrigin(sourceURL), "not_found").Inc()
+		level.Debug(store.l).Log("msg", "sourcemap not found on filesystem", "url", sourceURL, "file_path", mapFilePath)
+		return nil, "", nil
+	}
+	level.Debug(store.l).Log("msg", "sourcemap found on filesystem", "url", mapFilePath, "file_path", mapFilePath)
+
+	content, err = store.fileService.ReadFile(mapFilePath)
+	if err != nil {
+		store.metrics.fileReads.WithLabelValues(getOrigin(sourceURL), "error").Inc()
+	} else {
+		store.metrics.fileReads.WithLabelValues(getOrigin(sourceURL), "ok").Inc()
+	}
+	return content, sourceURL, err
+}
+
+func (store *RealSourceMapStore) getSourceMapContent(sourceURL string, release string) (content []byte, sourceMapURL string, err error) {
+	//attempt to find in fs
+	for _, fileconf := range store.config.FileSystem {
+		content, sourceMapURL, err = store.getSourceMapFromFileSystem(sourceURL, release, fileconf)
+		if content != nil || err != nil {
+			return content, sourceMapURL, err
+		}
+	}
+
+	//attempt to download
+	if strings.HasPrefix(sourceURL, "http") && utils.URLMatchesOrigins(sourceURL, store.config.DownloadFromOrigins) {
+		return store.downloadSourceMapContent(sourceURL)
+	}
+	return nil, "", nil
+}
+
+func (store *RealSourceMapStore) getSourceMap(sourceURL string, release string) (*sourceMap, error) {
+	store.Lock()
+	defer store.Unlock()
+
+	cacheKey := fmt.Sprintf("%s__%s", sourceURL, release)
+
+	if smap, ok := store.cache[cacheKey]; ok {
+		return smap, nil
+	}
+	content, sourceMapURL, err := store.getSourceMapContent(sourceURL, release)
+	if err != nil || content == nil {
+		store.cache[cacheKey] = nil
```

Comment:
Why is this needed? Are we terminating the sequence after that? If not, this can cause a potential memory leak. I suggest using the [delete](https://pkg.go.dev/builtin#delete) function to remove the key completely

Reply:
This is overwriting an existing item in the map, this is adding a not yet existing key with `nil` value so we don't try to download again if first time didn't work. So should be ok mem leak wise.

That said, accumulation of stuff in the cache might be a problem down the road that perhaps could be solved with adding some sort of mechanism to delete items that haven't been used for a while